### DISCUSSION TITLE
System.MissingMethodException when chocolatey.lib is used out of choco.exe

### DISCUSSION THREAD

**Discussion by zbalkan** [chocolatey/choco#2395] [Q&A] [answered]
### What You Are Seeing?
I am a user of [BulkCrapUninstaller](https://github.com/Klocman/Bulk-Crap-Uninstaller) and had problems with the amount of time the application wastes while querying Chocolatey package information. I wanted to create an issue on the said project but it turned out there is already an [open ticket](https://github.com/Klocman/Bulk-Crap-Uninstaller/issues/248). The software is called `choco.exe` with required parameters once for the list of packages and then one for each of the packages for detailed information. It is possible to check [the code](https://github.com/Klocman/Bulk-Crap-Uninstaller/blob/910d1fe82f656dc2ed0032401176d9739feb85fc/source/UninstallTools/Factory/ChocolateyFactory.cs#L68).
In the said issue, it was suggested to get rid of process creation overhead by running the `choco.exe` once with details and then parse the information. My solution was to utilize `chocolatey.lib` for the requirements.

I created a [new benchmark project](https://github.com/zbalkan/BenchmarkChoco) and did some benchmarks there. It has significant advantages over other methods -**600x faster**. However, when I added the code to the [fork of BulkCrapUninstaller](https://github.com/zbalkan/Bulk-Crap-Uninstaller) and run it, I get the error below:

```
System.MissingMethodException
Method not found: 'Void System.Threading.Mutex.SetAccessControl(System.Security.AccessControl.MutexSecurity)'.
   at chocolatey.infrastructure.synchronization.GlobalMutex.init_mutex()
   at chocolatey.infrastructure.synchronization.GlobalMutex..ctor(Int32 timeOut)
   at chocolatey.infrastructure.synchronization.GlobalMutex.enter[T](Func`1 func, Int32 timeout)
   at chocolatey.Lets.GetChocolatey(Boolean initializeLogging)
   at UninstallTools.Factory.ChocolateyFactory.GetUninstallerEntries(ListGenerationCallback progressCallback) in <path to repos>\Bulk-Crap-Uninstaller\Bulk-Crap-Uninstaller\source\UninstallTools\Factory\ChocolateyFactory.cs:line 32
   at UninstallTools.Factory.ApplicationUninstallerFactory.GetMiscUninstallerEntries(ListGenerationCallback progressCallback) in <path to repos>\Bulk-Crap-Uninstaller\Bulk-Crap-Uninstaller\source\UninstallTools\Factory\ApplicationUninstallerFactory.cs:line 267
```

### What is Expected?
No exception on running the code using the `chocolatey.lib`.

### How Did You Get This To Happen? (Steps to Reproduce)
- Clone the https://github.com/zbalkan/Bulk-Crap-Uninstaller repository.
- Ensure the dependencies are met (BCU-Launcher project is a C++ project that requires the latest MFC v142, others are written C#).
- Build and run the project.

**Comment by zbalkan** [comment]
I'd like to comment on a few stuff before an analysis of the issue:

1. I am aware that the `chocolatey.lib` may not be designed for other applications in mind. So, the issue may not be related to the value created by the chocolatey project.
2. Since it was not designed as an API for 3rd parties, there may be license checks and other controls in place which may affect the use of the library. This may be related and consistent for Chocolatey and ChocolateyGUI, but not 3rd parties.
3. For the same reason, there was no documentation about the library. Although it is easy to get the basics, it became harder to get into the Chocolatey source to find the probable cause when I had this issue.

**Comment by TheCakeIsNaOH** [comment, marked as answer]
I suspect that [this](https://github.com/Klocman/Bulk-Crap-Uninstaller/blob/910d1fe82f656dc2ed0032401176d9739feb85fc/source/UninstallTools/UninstallTools.csproj#L3) is your problem. The `UninstallTools` project is running on .Net 5. Chocolatey is built on .Net 4.0 

`System.Threading.Mutex.SetAccessControl` is available in .Net framework, but not available in .Net core/5, at least not in the same place.

**Reply by gep13** [reply]
@zbalkan did the comment from @TheCakeIsNaOH make sense here? Have you been able to progress this any further?

In the future, can I ask that this type of question is asked within the Discussions tab? We can always create an issue from the Discussion if required.

For now, I will move this into the Discussion area.

**Reply by zbalkan** [reply]
Thank you for the action. I asked this one in gitter first, but this might be a better place.

And @TheCakeIsNaOH is right. It's the target frameworks. And my PR there will not be applicable until chocolatey.lib targets dotnet 5.

**Reply by TheCakeIsNaOH** [reply]
@zbalkan I did a bit more research, and you may be able to use [Windows Compatibility Pack for .NET Core](https://docs.microsoft.com/en-us/archive/msdn-magazine/2017/connect/net-introducing-the-windows-compatibility-pack-for-net-core). It has `System.Threading.AccessControl` available in the pack, so it may work,

**Reply by zbalkan** [reply]
I was thinking of using a hot patching library like [Harmony](https://github.com/pardeike/Harmony) so that I would override the related methods.

To clarify the Compatibility Pack thing, you are suggesting to modify chocolatey.lib, right? Because the problem occurs when I use the `Lets.GetChocolatey()` method and I am not sure how this pack would help me. Since I do not know Chocolaty internals, it's a little bit ambiguous to me. Can you please elaborate on your suggestion?

**Reply by TheCakeIsNaOH** [reply]
I haven't used it before, so I'm not quite sure how it is implemented in practice. 

Using a hot patching library like you suggested might be a good option.

**Reply by zbalkan** [reply]
Do you have any idea if or when chocolatey will change the target framework? It is it OK to send a PR for multitargeting?

**Reply by gep13** [reply]
@zbalkan this is on our internal roadmap for Chocolatey, but this is not something that we are looking at just now.  There are lots of moving parts to Chocolatey which would be impacted by this type of change, i.e. Chocolatey Licensed Extension, Chocolatey GUI, Chocolatey Agent, etc.  So, no, at this time, a PR attempting to change target frameworks couldn't be accepted.

**Reply by zbalkan** [reply]
@gep13 Thank you for the information. I'll not interfere with an open source project's roadmap to contribute another one. I guess I'll wait for some time. Thank you again for the details.
